server/metadata/queue.go

```diff
@@ -0,0 +1,241 @@
+}
+
+func NewQueueItem(priority int64, data []byte) *QueueItem {
+	return &QueueItem{
+		Id:         uuid.NewUUIDAsString(),
+		Priority:   priority,
+		ErrorCount: 0,
+		Data:       data,
+	}
+}
+
+func newQueueStore(nameRegistry *NameRegistry) *QueueSubspace {
+	return &QueueSubspace{
+		metadataSubspace{
+			SubspaceName: nameRegistry.QueueSubspaceName(),
+			KeyVersion:   []byte{byte(queueMetaValueVersion)},
+		},
+	}
+}
+
+func (q *QueueSubspace) getKey(vestingTime int64, priority int64, id string) keys.Key {
+	if id == maxId {
+		return keys.NewKey(q.SubspaceName, q.KeyVersion, itemsSpace, vestingTime, priority, 0xFF)
+	}
+
+	return keys.NewKey(q.SubspaceName, q.KeyVersion, itemsSpace, vestingTime, priority, id)
+}
+
+func (q *QueueSubspace) Enqueue(ctx context.Context, tx transaction.Tx, item *QueueItem, delay time.Duration) error {
+	if len(item.Id) == 0 {
+		return errors.Internal("cannot enqueue without an id")
+	}
+	item.VestingTime = time.Now().Add(delay).UnixMilli()
+	key := q.getKey(item.VestingTime, item.Priority, item.Id)
+	tableData, err := q.encodeItem(item)
+	if err != nil {
+		return err
+	}
+	return tx.Replace(ctx, key, tableData, false)
+}
+
+func (q *QueueSubspace) Peak(ctx context.Context, tx transaction.Tx, max int) ([]QueueItem, error) {
+	var items []QueueItem
+	count := 0
+	currentTime := time.Now().UnixMilli()
+	endKey := q.getKey(currentTime, maxPriority, maxId)
+	iter, err := tx.ReadRange(ctx, nil, endKey, false)
+	if err != nil {
+		return items, err
+	}
+	var v kv.KeyValue
+	for iter.Next(&v) && count < max {
+		count += 1
+		item, err := q.decodeItem(v.Data)
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, *item)
+	}
+
+	return items, iter.Err()
+}
+
+func (q *QueueSubspace) Find(ctx context.Context, tx transaction.Tx, item *QueueItem) (*QueueItem, error) {
+	startKey := q.getKey(item.VestingTime, 0, item.Id)
+	endKey := q.getKey(item.VestingTime, maxPriority, item.Id)
+
+	iter, err := tx.ReadRange(ctx, startKey, endKey, false)
+	if err != nil {
+		return nil, err
+	}
+	var v kv.KeyValue
+	for iter.Next(&v) {
+		found, err := q.decodeItem(v.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		if item.Id == found.Id {
+			return item, nil
+		}
+	}
+
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+	return nil, errors.Internal("QueueItem \"%s\" was not found", item.Id)
+}
+
+func (q *QueueSubspace) Dequeue(ctx context.Context, tx transaction.Tx, item *QueueItem) error {
+	itemKey := q.getKey(item.VestingTime, item.Priority, item.Id)
+	return tx.Delete(ctx, itemKey)
+}
+
+func (q *QueueSubspace) ObtainLease(ctx context.Context, tx transaction.Tx, item *QueueItem, leaseTime time.Duration) (*QueueItem, error) {
+	item, err := q.Find(ctx, tx, item)
+	if err != nil {
+		return nil, err
+	}
+	if item == nil {
+		return nil, errors.Internal("Queue Item was not found")
+	}
+
+	if err = q.Dequeue(ctx, tx, item); err != nil {
+		return nil, err
+	}
+
+	item.LeaseId = uuid.NewUUIDAsString()
```

Comment:
Ideally this should be passed by the owner?

Reply:
I would like to leave as is for now. I might change it to the owner once I implement the workers part.

Reply:
Okay - the other thing is that as it is the same item and must be the same worker renewing the lease do we need to change the lease id?

Reply:
No the lease id should stay the same while a worker works on it.

Reply:
I thought it is renewLease, nvm.